logic/scanning_logic.py

```diff
@@ -0,0 +1,629 @@
+                    y_index=self.__scan_line_count-1)
+            else:
+                scan_point = self._current_dummy_data[self.__scan_line_count - 1]
+                self._scan_data[self.__running_scan].add_data_point(
+                    position=self.__scan_line_positions,
+                    data={chnl: scan_point * (i+1) for i, chnl in enumerate(channels)},
+                    index=self.__scan_line_count-1)
+
+            self.sigScanDataChanged.emit({self.__running_scan: self.scan_data[self.__running_scan]})
+            self.__sigNextLine.emit()
+        return
+
+    def _generate_2d_dummy_data(self, axes):
+        x_res = self._scanner_settings['scan_resolution'][axes[0]]
+        y_res = self._scanner_settings['scan_resolution'][axes[1]]
+        x_start = self._scanner_settings['scan_range'][axes[0]][0]
+        y_start = self._scanner_settings['scan_range'][axes[1]][0]
+        z_start = -5e-6
+        x_end = self._scanner_settings['scan_range'][axes[0]][1]
+        y_end = self._scanner_settings['scan_range'][axes[1]][1]
+        z_end = 5e-6
+        x_range = x_end - x_start
+        y_range = y_end - y_start
+        z_range = z_end - z_start
+
+        area_density = 1 / (5e-6 * 5e-6)
+
+        params = np.random.rand(round(area_density * x_range * y_range), 7)
+        params[:, 0] = params[:, 0] * x_range + x_start     # X displacement
+        params[:, 1] = params[:, 1] * y_range + y_start     # Y displacement
+        params[:, 2] = params[:, 2] * z_range + z_start     # Z displacement
+        params[:, 3] = params[:, 3] * 50e-9 + 150e-9        # X sigma
+        params[:, 4] = params[:, 4] * 50e-9 + 150e-9        # Y sigma
+        params[:, 5] = params[:, 5] * 100e-9 + 450e-9       # Z sigma
+        params[:, 6] = params[:, 6] * 2 * np.pi             # theta
+
+        amplitude = 200000
+        offset = 20000
+
+        def gauss_ensemble(x, y):
+            result = np.zeros(x.shape)
+            for x0, y0, z0, sigmax, sigmay, sigmaz, theta in params:
+                a = np.cos(theta) ** 2 / (2 * sigmax ** 2) + np.sin(theta) ** 2 / (2 * sigmay ** 2)
+                b = np.sin(2 * theta) / (4 * sigmay ** 2) - np.sin(2 * theta) / (4 * sigmax ** 2)
+                c = np.sin(theta) ** 2 / (2 * sigmax ** 2) + np.cos(theta) ** 2 / (2 * sigmay ** 2)
+                zfactor = np.exp(-(z0 ** 2) / (2 * sigmaz**2))
+                result += zfactor * np.exp(
+                    -(a * (x - x0) ** 2 + 2 * b * (x - x0) * (y - y0) + c * (y - y0) ** 2))
+            result *= amplitude - offset
+            result += offset
+            return result
+
+        xx, yy = np.meshgrid(np.linspace(x_start, x_end, x_res),
+                             np.linspace(y_start, y_end, y_res),
+                             indexing='ij')
+        return np.random.rand(xx.shape[0], xx.shape[1]) * amplitude * 0.10 + gauss_ensemble(xx, yy)
+
+    def generate_1d_dummy_data(self, axis):
```

Comment:
why underscore the generate function of the 2d data but not the 1d data?

Reply:
Well, this logic module is just an intermediate solution to demonstrate the GUI functionality. This is not polished and never will be since it will be discarded in the near future. The only part that will probably be reused is the `ScanData` class.